### PR TITLE
CXX-2762 Ensure ABI namespace is included in API documentation

### PR DESCRIPTION
**Pull request by eramongodb**
Resolves CXX-2762. Verified by [this patch](https://spruce.mongodb.com/version/652d8da12fbabed3205c7cd0).

## Description

Despite the diffs to source code, this PR is primarily focused on changes to generated documentation, not the source code itself. Most of the diff is due to replacing use of the inline namespace macros with `inline namespace v_noabi`. This is to allow Doxygen to understand the declaration scope of entities within the inline namespace, as it is incapable of reading into the expansion of the inline namespace macro, and Doxygen does not provide a means to manually open/close a namespace for documentation purposes. Explicitly substituting the macros appears to be the only method for ensuring namespace-awareness in documentation.

To elaborate, using the `mongocxx::instance` class as an example, the current API documentation generates the [class list](http://mongocxx.org/api/current/annotated.html) as follows:
```
mongocxx
├── instance
└── ...
```
And the [class reference page](http://mongocxx.org/api/current/classmongocxx_1_1instance.html) has no awareness of its actual declaration within the `v_noabi` inline namespace:
```
mongocxx::instance Class Reference
```

This is an issue if we want to add documentation for upcoming `v1` namespace entities in a manner that distinguishes `v1` entities from `v_noabi` entities, in particular if we want to support a mix of `vN` entities in the root namespace. The result of the changes in this PR changes the class index to:
```
mongocxx
├── v_noabi
│   ├── instance
│   └── ...
├── instance (v_noabi/instance)
└── ...
```
And the class reference to:
```
mongocxx::v_noabi::instance Class Reference
```

The `instance` class appears twice in the class index to reflect inline namespace behavior (`mongocxx::instance` _and_ `mongocxx::v_noabi::instance`). The duplication of entities in the lists under the root and `v_noabi` namespace may seem redundant, but this is correct behavior: the root namespace declaration may redirect to any `vN` namespace declaration instead (e.g. `mongocxx::instance -> mongocxx::v1::instance`) depending on the state of compatibility with prior ABI versions and/or the unstable ABI (i.e. eventual `v_noabi` -> `v1` source-but-not-binary-breaking changes). The changes in this PR enables the generated API documentation to make this distinction rather than tying the root namespace declaration to the entity in the currently inlined namespace (there is no actual `mongocxx::instance`; it's actually `mongocxx::v_noabi::instance`).

Ensuring proper inclusion of the ABI namespace in Doxygen required a definitive Doxygen comment for the enclosing namespaces. This is currently done in ad-hoc locations (json.hpp for the `bsoncxx` namespace, client.hpp for the `mongocxx` namespace). This PR moves the namespace documentation comments into the prelude headers for each respective library for uniformity.

## Drive-By Improvements

* Update the Doxyfile configuration to account for the directory restructure made in https://github.com/mongodb/mongo-cxx-driver/pull/1026. Note: only _non-generated_, _public_ header files are given as input to Doxygen. Implementation, private, and test files are ignored even if they still contain Doxygen comments (which can be useful for IDE-integration purposes even if they don't end up in generated documentation). This better reflects the intended input to Doxygen even prior to the directory restructure.
* Address various warnings currently generated by Doxygen, including:
  * Undocumented public classes (`chunks_and_bytes_offset`, `rewrap_many_datakey`).
  * Redundant/useless `@relates` in friend declarations.
  * Undocumented/misdocumented parameters.
  * Ambiguous/unclear friend declarations (missing class key and/or forward declaration).
  * Missing parameter documentation in grouped member functions in `mongocxx::gridfs::bucket`: used `@copydoc` instead to allow appending parameter documentation.
  * Missing `@code`/`@endcode` commands to wrap literal blocks (`<>` is incorrectly parsed as XML).

### CODE REVIEW

**Review by kevinAlbs — APPROVED** [review]
LGTM with removal of unused macros.

File: src/bsoncxx/lib/bsoncxx/v_noabi/bsoncxx/config/config.hpp.in [hunk 1]
```diff
@@ -19,5 +19,4 @@
 #cmakedefine BSONCXX_POLY_USE_STD
 
 #define BSONCXX_INLINE_NAMESPACE_BEGIN inline namespace v_noabi {
-
-#define BSONCXX_INLINE_NAMESPACE_END }  // namespace v_noabi
+#define BSONCXX_INLINE_NAMESPACE_END  // namespace v_noabi
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply src/bsoncxx/lib/bsoncxx/v_noabi/bsoncxx/config/config.hpp.in]: These macros are no longer used in CXX Driver code (and I do not intend to reuse/repurpose them in future PRs either), but they should remain for backward source compatibility. The removal of the closing brace was unintentional.

**Comment by eramongodb** [comment]
Resolved merge conflicts.

Also added some missing generated macros to the Doxyfile to address parsing issues for certain entities (i.e. misinterpreting typedefs as variable declarations).

**Comment by eramongodb** [comment]
To further account for changes in https://github.com/mongodb/mongo-cxx-driver/pull/1028, documentation for the `bsoncxx::v_noabi::stdx` namespace was added so that entities declared within (i.e. our own `make_unique()`) are included in generated documentation.

Additionally, exclusion filters for `detail` namespaces under `bsoncxx::` and `mongocxx::` were added to silence Doxygen warnings for undocumented symbols within (i.e. `make_unique_impl`).

**Review by kkloberdanz — APPROVED** [review]

**Pull request merged**